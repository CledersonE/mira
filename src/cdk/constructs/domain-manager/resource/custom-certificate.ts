import { Topic } from '@aws-cdk/aws-sns'
import { Construct } from '@aws-cdk/core'
import {
  CustomResource,
  CustomResourceProvider
} from '@aws-cdk/aws-cloudformation'
import { MiraConfig } from '../../../..'
import { CERTIFICATE_SUBSCRIPTION_TOPIC } from './certificate-manager'
import { CROSS_ACCOUNT_DOMAIN_MANAGER_ROLE } from './route53-manager'

export interface CustomCertificateProps {
  domain: string
}

export class CustomCertificate extends Construct {
  certificate: CustomResource
  certificateArn: string

  constructor (parent: Construct, props: CustomCertificateProps) {
    super(parent, 'CustomCertificateResouce')

    const topicName = MiraConfig.getBaseStackName(CERTIFICATE_SUBSCRIPTION_TOPIC)
    const accountConfig = MiraConfig.getEnvironment()
    const domainAccountConfig = MiraConfig.getEnvironment('domain')

    const CertificateProviderTopic = Topic.fromTopicArn(
      this,
      'CertificateProvider',
      `arn:aws:sns:${accountConfig.env.region}:${accountConfig.env.account}:${topicName}`
    )

    this.certificate = new CustomResource(this, 'CustomCertificate', {
      provider: CustomResourceProvider.fromTopic(
        CertificateProviderTopic
      ),
      properties: {
        Domain: props.domain,
        Route53Role: `arn:aws:iam::${domainAccountConfig.env.account}:role/${MiraConfig.getBaseStackName(CROSS_ACCOUNT_DOMAIN_MANAGER_ROLE)}`
      },
      resourceType: 'Custom::Certificate'
    })

    this.certificateArn = this.certificate.getAttString('arn')
  }
}
