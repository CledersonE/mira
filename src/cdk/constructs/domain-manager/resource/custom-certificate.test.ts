import { SynthUtils } from "@aws-cdk/assert"
import { Stack, App } from "@aws-cdk/core"

import {
  MiraConfig,
  DomainConfig,
  Account,
} from "../../../../config/mira-config"

import { CustomCertificate } from "./custom-certificate"

describe("CustomCertificate", () => {
  test("renders output correctly", () => {
    const stack = new Stack(
      new App(),
      MiraConfig.getBaseStackName("CustomCertificate"),
      {}
    )
    MiraConfig.getEnvironment = (envName: string): Account => {
      if (envName === 'domain') {
        return {
          name: "domain",
          profile: "domain",
          env: { account: "domain", region: "eu-west-1" },
        }
      }

      return {
        name: "account1",
        profile: "account1",
        env: { account: "account1", region: "eu-west-1" },
      }
    }

    MiraConfig.getDomainConfig = (): DomainConfig => ({
      hostedZoneId: 'Z123456789',
      accounts: [
        'account1',
        'domain'
      ],
    })

    const f = new CustomCertificate(stack, {
      domain: 'test.example.com'
    })
    expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot()
  })
})
