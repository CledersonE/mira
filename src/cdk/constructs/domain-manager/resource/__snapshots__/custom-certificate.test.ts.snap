// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CustomCertificate renders output correctly 1`] = `
Object {
  "Resources": Object {
    "CustomCertificateResourceCustomCertificateA17D7D34": Object {
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "Domain": "test.example.com",
        "Route53Role": "arn:aws:iam::domain:role/John-MyGreatApp-CrossAccountDomainManagerRole",
        "ServiceToken": "arn:aws:sns:eu-west-1:account1:John-MyGreatApp-CertificateSubscriptionTopic",
      },
      "Type": "Custom::Certificate",
      "UpdateReplacePolicy": "Delete",
    },
  },
}
`;
